### ISSUE TITLE
Make `Measure::this()` more ergonomic

### ISSUE THREAD

**Issue by brooksprumo** [solana-labs/solana#23084]
## Problem

[`Measure::this()`](https://github.com/solana-labs/solana/blob/c078ca3fb3e203dbd4f3a9556166695e8223cea7/measure/src/measure.rs#L42-L85) is useful for easily adding timing around a function or a block. However, it is not ergonomic.  

`Measure::this()` is not so bad when the function takes only one parameter. But with zero or 2+ parameters, the long-form closure syntax is required. Similar workarounds are required when calling functions that take `self` (i.e. methods in Impl blocks). See below for examples.

Usage with zero-parameter functions in bank.rs:
https://github.com/solana-labs/solana/blob/c078ca3fb3e203dbd4f3a9556166695e8223cea7/runtime/src/bank.rs#L1575-L1625

Usage in test for `Measure::this()`:
https://github.com/solana-labs/solana/blob/c078ca3fb3e203dbd4f3a9556166695e8223cea7/measure/src/measure.rs#L176-L239

## Proposed Solution

Write a macro (maybe named `measure!()`) that handles calling `Measure::this()` (or performs similar functionality) on 0, 1, and 2+ parameter functions (and blocks).

Consider [the `dbg!()` macro](https://doc.rust-lang.org/std/macro.dbg.html); it just wraps existing expressions. It would be prudent to model `measure_this!()` after `dbg!()`.

For example:
```rust

fn zero() { ... };
fn one(x: u64) { ... };
fn two (x: u32, y: u32) { ... };

struct Foo { ... }
impl Foo {
    fn bar() { ... };
    fn baz(a: &str, b: &str, c: &str) { ... };
}

fn test {
    // functions
    let (result, measure) = measure!(zero(), "zero params");
    let (result, measure) = measure!(one(42), "one param");
    let (result, measure) = measure!(two(11, 22), "two params");

    // methods
    let foo = Foo { ... };
    let (result, measure) = measure!(foo.bar(), "method without params");
    let (result, measure) = measure!(foo.baz("a", "b", "c"), "method with params");

    // blocks
    let (result, measure) = measure!({
            // ... perform some involved and/or time consuming computation ...
            std::thread::sleep(std::time::Duration::from_secs(10));
            1 + 2 
        },
        "some involved block"
    );
}
```

**Comment by kwsohn** [comment]
@brooksprumo Hey Brooks, Is this being actively worked on? If not, I'd be happy to take this on.

**Comment by brooksprumo** [comment]
@kwsohn Please do, thanks!

**Comment by brooksprumo** [comment]
Fixed by #24137
